File: leetpattern/cpp/0001_0300/0290_word_pattern.cc
#include <cassert>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

bool wordPattern(const string& pattern, const string& s) {
    vector<string> words;
    istringstream iss(s);
    string word;
    while (iss >> word) {
        words.push_back(word);
    }
    if (pattern.size() != words.size()) return false;

    unordered_map<char, string> p2w;
    unordered_map<string, char> w2p;

    for (size_t i = 0; i < pattern.size(); ++i) {
        char p = pattern[i];
        const string& w = words[i];
        if ((p2w.count(p) && p2w[p] != w) || (w2p.count(w) && w2p[w] != p)) {
            return false;
        }
        p2w[p] = w;
        w2p[w] = p;
    }
    return true;
}

int main() {
    assert(wordPattern("abba", "dog cat cat dog") == true);
    assert(wordPattern("abba", "dog cat cat fish") == false);
    assert(wordPattern("aaaa", "dog cat cat dog") == false);
    assert(wordPattern("abba", "dog dog dog dog") == false);
    assert(wordPattern("abc", "b c a") == true);
    assert(wordPattern("abc", "b c a a") == false);
    assert(wordPattern("ab", "b b") == false);
    cout << "All test cases pass" << endl;
    return 0;
}

File: leetpattern/python/0001_0300/0290_word_pattern.py
def wordPattern(pattern: str, s: str) -> bool:
    words = s.split()
    if len(pattern) != len(words):
        return False

    p2w, w2p = {}, {}
    for p, w in zip(pattern, words):
        if (p in p2w and p2w[p] != w) or (w in w2p and w2p[w] != p):
            return False
        p2w[p] = w
        w2p[w] = p
    return True


def test_wordPattern():
    assert wordPattern("abba", "dog cat cat dog")
    assert not wordPattern("abba", "dog cat cat fish")
    assert not wordPattern("aaaa", "dog cat cat dog")
    assert not wordPattern("abba", "dog dog dog dog")
    assert wordPattern("abc", "b c a")
    assert not wordPattern("abc", "b c a a")
    assert not wordPattern("ab", "b b")
